File: 백준/Silver/1697. 숨바꼭질/숨바꼭질.py
import sys
from collections import deque

# 3가지 이동 방법이 있음
# x-1, x+1, x*2

def bfs():
    global N, K, visited
    q = deque()
    q.append((N, 0))
    while q:
        temp, count = q.popleft()
        for i in range(3):
            next_temp = temp
            if i == 0 and temp < 100000 :
                next_temp += 1
            elif i == 1 and temp >= 1:
                next_temp -= 1
            elif i == 2 and temp <= 50000:
                next_temp *= 2
            else:
                continue
                
            if visited[next_temp] == 0:
                visited[next_temp] = count + 1
                q.append((next_temp, count + 1))

                if next_temp == K:
                    break



N, K = map(int, sys.stdin.readline().split())
visited = [0 for _ in range(100001)]

visited[N] = 1

if N == K:
    print(0)
else:
    bfs()
    print(visited[K])

File: 백준/Silver/1697. 숨바꼭질/숨바꼭질.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;

public class Main {

	static int N, K = 0;
	static boolean[] visited = new boolean[100001];
	static int[] arr = new int[100001];

	public static void main(String[] args) throws IOException {
		BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));

		String[] sa = bf.readLine().split(" ");
		N = Integer.parseInt(sa[0]);
		K = Integer.parseInt(sa[1]);

		if (N == K) {
			System.out.println(0);
		} else {
			int result = bfs();
			System.out.println(result);
		}
	}

	public static int bfs() {
		Queue<int[]> q = new LinkedList<>();
		q.add(new int[] {N, 0});
		visited[N] = true;

		while (!q.isEmpty()) {
			int[] temp = q.poll();
			int cur = temp[0];
			int second = temp[1];

			if (cur == K)
				return second;

			if (cur - 1 >= 0 && !visited[cur - 1]) {
				visited[cur - 1] = true;
				q.add(new int[] {cur - 1, second + 1});
			}

			if (cur + 1 <= 100000 && !visited[cur + 1]) {
				visited[cur + 1] = true;
				q.add(new int[] {cur + 1, second + 1});
			}

			if (cur * 2 <= 100000 && !visited[cur * 2]) {
				visited[cur * 2] = true;
				q.add(new int[] {cur * 2, second + 1});
			}
		}

		return 0;
	}
}